_posts/2024-04-20-websockets2.md

```diff
@@ -0,0 +1,1861 @@
+
+```scala
+package rockthejvm.websockets.codecs
+
+import skunk.*
+import skunk.codec.all.*
+import rockthejvm.websockets.domain.user.{UserId, UserName}
+import rockthejvm.websockets.domain.room.{RoomId, RoomName}
+import rockthejvm.websockets.domain.message.{
+  MessageId,
+  MessageText,
+  MessageTime
+}
+
+object codecs {
+  val userId: Codec[UserId] = uuid.imap[UserId](UserId(_))(_.id)
+  val userName: Codec[UserName] =
+    varchar(255).imap[UserName](UserName(_))(_.name)
+  val roomId: Codec[RoomId] = uuid.imap[RoomId](RoomId(_))(_.id)
+  val roomName: Codec[RoomName] =
+    varchar(255).imap[RoomName](RoomName(_))(_.name)
+  val messageId: Codec[MessageId] = uuid.imap[MessageId](MessageId(_))(_.id)
+  val messageText: Codec[MessageText] =
+    text.imap[MessageText](MessageText(_))(_.value)
+  val messageTime: Codec[MessageTime] =
+    timestamp.imap[MessageTime](MessageTime(_))(_.time)
+}
+```
+Skunk provides several important codecs through the `skunk.codec.all.*` import such as `uuid`, `varchar`, `text`, and `timestamp`. We use the `imap` method to provide the `encode` and `decode` methods for each case class, here's the signature:
+
+```scala
+  def imap[B](f: A => B)(g: B => A): Codec[B] =
+    Codec(b => encode(g(b)), decode(_, _).map(f), types)
+```
+Here we provide the methods from case class to postgres and the reverse. Here we provide codecs for case classes that take native types, however, later we shall see how to join codecs for types such as `InsertMessage`.
+
+### 5.1 The Postgres Protocols
+
+In this section, we'll create various methods that will interact with the Postgres database through `Skunk`. Let's create a `PostgresProtocol.scala` file under the `websockets` and add the following code:
+
+```scala
+package rockthejvm.websockets
+
+import fs2.Stream
+import java.time.LocalDateTime
+import rockthejvm.websockets.domain.user.*
+import rockthejvm.websockets.domain.room.*
+import rockthejvm.websockets.domain.message.*
+
+trait PostgresProtocol[F[_]] {
+  def createUser(name: String): F[Either[String, User]]
+  def createRoom(name: String): F[Either[String, Room]]
+  def deleteRoom(roomId: RoomId): F[Unit]
+  def deleteUser(userId: UserId): F[Unit]
+  def saveMessage(
+      message: String,
+      time: LocalDateTime,
+      userId: UserId,
+      roomId: RoomId
+  ): F[Unit]
+  def fetchMessages(roomId: RoomId): F[Stream[F, FetchMessage]]
+  def deleteRoomMessages(roomId: RoomId): F[Unit]
+}
+```
+First, we create a `PostgresProtocol` trait with a number of methods that will later be used by our chat protocol:
+
+```scala
+...
+import cats.effect.std.UUIDGen
+import cats.effect.* 
+import skunk.* 
+
+trait PostgresProtocol[F[_]] {
+  ...
+}
+
+object PostgresProtocol {
+  def make[F[_]: UUIDGen: MonadCancelThrow: Concurrent](
```

Comment:
`MonadCancelThrow` shouldn't be needed here, Concurrent is already more powerful